In [58]:
# Use for Python 3.9
!pip3.9 install -r sweetviz_requirements.txt

DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621
DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621


In [1]:
# Use for Python 3.11
!pip3.11 install -r ydata_requirements.txt


[notice] A new release of pip is available: 24.3.1 -> 25.0.1
[notice] To update, run: python3.11 -m pip install --upgrade pip


In [5]:
import pandas as pd
import numpy as np
from ydata_profiling import ProfileReport

In [6]:
# noinspection PyPackageRequirements
print(pd.__version__)
print(np.__version__)

2.2.2
2.1.3


In [7]:
df = pd.read_csv('expanded_data_again.csv')
df.head()

Index                   Timestamp  Feed timer Last Cycle Time  \
0      4  2025-03-03 15:11:06.163873    694348.0     00000:02:39   
1      5  2025-03-03 15:11:06.209764    694348.0     00000:02:39   
2      6  2025-03-03 15:11:06.335765    694348.0     00000:02:39   
3      7  2025-03-03 15:11:07.035919    694348.0     00000:02:39   
4      8  2025-03-03 15:11:07.160275    694348.0     00000:02:39   

  Year month day  Cycle start timer  Tool Number in use  Max axis load for X  \
0     2025/03/03           986332.0                  10                  0.0   
1     2025/03/03           986332.0                  10                 -1.0   
2     2025/03/03           986332.0                  10                  0.0   
3     2025/03/03           986332.0                  10                 -1.0   
4     2025/03/03           986332.0                  10                  0.0   

   Max axis load for Y  Max axis load for Z  ...  \
0                 23.0                  0.0  ...   
1                 23.0                 -1.0  ...   
2                 23.0                  0.0  ...   
3                 23.0                 -1.0  ...   
4                 23.0                  0.0  ...   

  Present part timer (read only) Previous part timer (read only)  \
0                          159.0                           159.0   
1                          159.0                           159.0   
2                          159.0                           159.0   
3                          159.0                           159.0   
4                          159.0                           159.0   

  Last Block (Modal) Group Codes 1  Last Block (Modal) Group Codes 2  \
0                              0.0                              17.0   
1                              0.0                              17.0   
2                              0.0                              17.0   
3                              0.0                              17.0   
4                              0.0                              17.0   

  Max recorded vibrations of tool 10  Last complete part timer (read only)  \
0                           1.519993                                 159.0   
1                           1.519993                                 159.0   
2                           1.519993                                 159.0   
3                           1.519993                                 159.0   
4                           1.519993                                 159.0   

   M30 Parts Counter #1 (resettable at control)  \
0                                          2101   
1                                          2101   
2                                          2101   
3                                          2101   
4                                          2101   

  M30 Parts Counter #2 (resettable at control)  \
0                                         2101   
1                                         2101   
2                                         2101   
3                                         2101   
4                                         2101   

   Spindle load with Haas vector drive (read only)  \
0                                              0.0   
1                                              0.0   
2                                              0.0   
3                                              0.0   
4                                              0.0   

   Three-in-one (PROGRAM Oxxxxx STATUS PARTS xxxxx)  
0                   PROGRAM, MDI, IDLE, PARTS, 2101  
1                   PROGRAM, MDI, IDLE, PARTS, 2101  
2                   PROGRAM, MDI, IDLE, PARTS, 2101  
3                   PROGRAM, MDI, IDLE, PARTS, 2101  
4                   PROGRAM, MDI, IDLE, PARTS, 2101  

[5 rows x 29 columns]

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19383 entries, 0 to 19382
Data columns (total 29 columns):
 #   Column                                            Non-Null Count  Dtype  
---  ------                                            --------------  -----  
 0   Index                                             19383 non-null  int64  
 1   Timestamp                                         19383 non-null  object 
 2   Feed timer                                        19383 non-null  float64
 3   Last Cycle Time                                   19383 non-null  object 
 4   Year month day                                    19383 non-null  object 
 5   Cycle start timer                                 19383 non-null  float64
 6   Tool Number in use                                19383 non-null  int64  
 7   Max axis load for X                               19383 non-null  float64
 8   Max axis load for Y                               19383 non-null  float64
 9   Max axis load for

In [9]:
df.describe()

Index  Feed timer  Cycle start timer  Tool Number in use  \
count  19383.00000     19383.0       19383.000000             19383.0   
mean    9695.00000    694348.0      986333.906206                10.0   
std     5595.53447         0.0           1.443989                 0.0   
min        4.00000    694348.0      986332.000000                10.0   
25%     4849.50000    694348.0      986332.000000                10.0   
50%     9695.00000    694348.0      986335.000000                10.0   
75%    14540.50000    694348.0      986335.000000                10.0   
max    19386.00000    694348.0      986335.000000                10.0   

       Max axis load for X  Max axis load for Y  Max axis load for Z  \
count         19383.000000         19383.000000         19383.000000   
mean             -0.499819            40.155858            -0.499716   
std               0.500013            12.995900             0.500013   
min              -1.000000            23.000000            -1.000000   
25%              -1.000000            23.000000            -1.000000   
50%               0.000000            50.000000             0.000000   
75%               0.000000            50.000000             0.000000   
max               0.000000            50.000000             0.000000   

       Tool Changes (total)  Spindle RPM (read only)  \
count               19383.0                  19383.0   
mean                 2975.0                      0.0   
std                     0.0                      0.0   
min                  2975.0                      0.0   
25%                  2975.0                      0.0   
50%                  2975.0                      0.0   
75%                  2975.0                      0.0   
max                  2975.0                      0.0   

       Power on timer (read only)  Tool in spindle (read only)  \
count                1.938300e+04                      19383.0   
mean                 1.590470e+07                         10.0   
std                  4.783108e+04                          0.0   
min                  1.583994e+07                         10.0   
25%                  1.584212e+07                         10.0   
50%                  1.593958e+07                         10.0   
75%                  1.594141e+07                         10.0   
max                  1.594326e+07                         10.0   

       Present part timer (read only)  Previous part timer (read only)  \
count                         19383.0                          19383.0   
mean                            159.0                            159.0   
std                               0.0                              0.0   
min                             159.0                            159.0   
25%                             159.0                            159.0   
50%                             159.0                            159.0   
75%                             159.0                            159.0   
max                             159.0                            159.0   

       Last Block (Modal) Group Codes 1  Last Block (Modal) Group Codes 2  \
count                           19383.0                           19383.0   
mean                                0.0                              17.0   
std                                 0.0                               0.0   
min                                 0.0                              17.0   
25%                                 0.0                              17.0   
50%                                 0.0                              17.0   
75%                                 0.0                              17.0   
max                                 0.0                              17.0   

       Max recorded vibrations of tool 10  \
count                        19383.000000   
mean                             1.519993   
std                              0.000000   
min                              1.519993   
25%                 

In [10]:
df.head(5)

Index                   Timestamp  Feed timer Last Cycle Time  \
0      4  2025-03-03 15:11:06.163873    694348.0     00000:02:39   
1      5  2025-03-03 15:11:06.209764    694348.0     00000:02:39   
2      6  2025-03-03 15:11:06.335765    694348.0     00000:02:39   
3      7  2025-03-03 15:11:07.035919    694348.0     00000:02:39   
4      8  2025-03-03 15:11:07.160275    694348.0     00000:02:39   

  Year month day  Cycle start timer  Tool Number in use  Max axis load for X  \
0     2025/03/03           986332.0                  10                  0.0   
1     2025/03/03           986332.0                  10                 -1.0   
2     2025/03/03           986332.0                  10                  0.0   
3     2025/03/03           986332.0                  10                 -1.0   
4     2025/03/03           986332.0                  10                  0.0   

   Max axis load for Y  Max axis load for Z  ...  \
0                 23.0                  0.0  ...   
1                 23.0                 -1.0  ...   
2                 23.0                  0.0  ...   
3                 23.0                 -1.0  ...   
4                 23.0                  0.0  ...   

  Present part timer (read only) Previous part timer (read only)  \
0                          159.0                           159.0   
1                          159.0                           159.0   
2                          159.0                           159.0   
3                          159.0                           159.0   
4                          159.0                           159.0   

  Last Block (Modal) Group Codes 1  Last Block (Modal) Group Codes 2  \
0                              0.0                              17.0   
1                              0.0                              17.0   
2                              0.0                              17.0   
3                              0.0                              17.0   
4                              0.0                              17.0   

  Max recorded vibrations of tool 10  Last complete part timer (read only)  \
0                           1.519993                                 159.0   
1                           1.519993                                 159.0   
2                           1.519993                                 159.0   
3                           1.519993                                 159.0   
4                           1.519993                                 159.0   

   M30 Parts Counter #1 (resettable at control)  \
0                                          2101   
1                                          2101   
2                                          2101   
3                                          2101   
4                                          2101   

  M30 Parts Counter #2 (resettable at control)  \
0                                         2101   
1                                         2101   
2                                         2101   
3                                         2101   
4                                         2101   

   Spindle load with Haas vector drive (read only)  \
0                                              0.0   
1                                              0.0   
2                                              0.0   
3                                              0.0   
4                                              0.0   

   Three-in-one (PROGRAM Oxxxxx STATUS PARTS xxxxx)  
0                   PROGRAM, MDI, IDLE, PARTS, 2101  
1                   PROGRAM, MDI, IDLE, PARTS, 2101  
2                   PROGRAM, MDI, IDLE, PARTS, 2101  
3                   PROGRAM, MDI, IDLE, PARTS, 2101  
4                   PROGRAM, MDI, IDLE, PARTS, 2101  

[5 rows x 29 columns]

## Data Preprocessing

In [11]:
mod_df = df.copy()
mod_df.head()

Index                   Timestamp  Feed timer Last Cycle Time  \
0      4  2025-03-03 15:11:06.163873    694348.0     00000:02:39   
1      5  2025-03-03 15:11:06.209764    694348.0     00000:02:39   
2      6  2025-03-03 15:11:06.335765    694348.0     00000:02:39   
3      7  2025-03-03 15:11:07.035919    694348.0     00000:02:39   
4      8  2025-03-03 15:11:07.160275    694348.0     00000:02:39   

  Year month day  Cycle start timer  Tool Number in use  Max axis load for X  \
0     2025/03/03           986332.0                  10                  0.0   
1     2025/03/03           986332.0                  10                 -1.0   
2     2025/03/03           986332.0                  10                  0.0   
3     2025/03/03           986332.0                  10                 -1.0   
4     2025/03/03           986332.0                  10                  0.0   

   Max axis load for Y  Max axis load for Z  ...  \
0                 23.0                  0.0  ...   
1                 23.0                 -1.0  ...   
2                 23.0                  0.0  ...   
3                 23.0                 -1.0  ...   
4                 23.0                  0.0  ...   

  Present part timer (read only) Previous part timer (read only)  \
0                          159.0                           159.0   
1                          159.0                           159.0   
2                          159.0                           159.0   
3                          159.0                           159.0   
4                          159.0                           159.0   

  Last Block (Modal) Group Codes 1  Last Block (Modal) Group Codes 2  \
0                              0.0                              17.0   
1                              0.0                              17.0   
2                              0.0                              17.0   
3                              0.0                              17.0   
4                              0.0                              17.0   

  Max recorded vibrations of tool 10  Last complete part timer (read only)  \
0                           1.519993                                 159.0   
1                           1.519993                                 159.0   
2                           1.519993                                 159.0   
3                           1.519993                                 159.0   
4                           1.519993                                 159.0   

   M30 Parts Counter #1 (resettable at control)  \
0                                          2101   
1                                          2101   
2                                          2101   
3                                          2101   
4                                          2101   

  M30 Parts Counter #2 (resettable at control)  \
0                                         2101   
1                                         2101   
2                                         2101   
3                                         2101   
4                                         2101   

   Spindle load with Haas vector drive (read only)  \
0                                              0.0   
1                                              0.0   
2                                              0.0   
3                                              0.0   
4                                              0.0   

   Three-in-one (PROGRAM Oxxxxx STATUS PARTS xxxxx)  
0                   PROGRAM, MDI, IDLE, PARTS, 2101  
1                   PROGRAM, MDI, IDLE, PARTS, 2101  
2                   PROGRAM, MDI, IDLE, PARTS, 2101  
3                   PROGRAM, MDI, IDLE, PARTS, 2101  
4                   PROGRAM, MDI, IDLE, PARTS, 2101  

[5 rows x 29 columns]

In [12]:
df['Index'].value_counts()

Index
4        1
12932    1
12930    1
12929    1
12928    1
        ..
6465     1
6464     1
6463     1
6462     1
19386    1
Name: count, Length: 19383, dtype: int64

In [13]:
mod_df.drop(columns=['Index'], inplace=True)

In [14]:
type(df['Timestamp'][0])

str

In [16]:
mod_df['Timestamp'] = pd.to_datetime(mod_df['Timestamp'])
type(mod_df['Timestamp'][0])

pandas._libs.tslibs.timestamps.Timestamp

In [17]:
df["Feed timer"].value_counts()

Feed timer
694348.0    19383
Name: count, dtype: int64

In [19]:
df["Last Cycle Time"].value_counts()

Last Cycle Time
00000:02:39    19383
Name: count, dtype: int64

In [20]:
df["Cycle start timer"].value_counts()

Cycle start timer
986335.0    12316
986332.0     7067
Name: count, dtype: int64

In [21]:
df["Tool Number in use"].value_counts()

Tool Number in use
10    19383
Name: count, dtype: int64

Feed timer: 694348, Last Cycle Time: 02:39, Tool Number in use: 10

In [22]:
mod_df.drop(columns=["Feed timer", "Last Cycle Time", "Tool Number in use"], inplace=True)
mod_df

Timestamp Year month day  Cycle start timer  \
0     2025-03-03 15:11:06.163873     2025/03/03           986332.0   
1     2025-03-03 15:11:06.209764     2025/03/03           986332.0   
2     2025-03-03 15:11:06.335765     2025/03/03           986332.0   
3     2025-03-03 15:11:07.035919     2025/03/03           986332.0   
4     2025-03-03 15:11:07.160275     2025/03/03           986332.0   
...                          ...            ...                ...   
19378 2025-03-05 13:23:45.301738     2025/03/05           986335.0   
19379 2025-03-05 13:23:45.429318     2025/03/05           986335.0   
19380 2025-03-05 13:23:45.561793     2025/03/05           986335.0   
19381 2025-03-05 13:23:46.174286     2025/03/05           986335.0   
19382 2025-03-05 13:23:46.439445     2025/03/05           986335.0   

       Max axis load for X  Max axis load for Y  Max axis load for Z  \
0                      0.0                 23.0                  0.0   
1                     -1.0                 23.0                 -1.0   
2                      0.0                 23.0                  0.0   
3                     -1.0                 23.0                 -1.0   
4                      0.0                 23.0                  0.0   
...                    ...                  ...                  ...   
19378                  0.0                 50.0                  0.0   
19379                 -1.0                 50.0                 -1.0   
19380                  0.0                 50.0                  0.0   
19381                 -1.0                 50.0                 -1.0   
19382                  0.0                 50.0                  0.0   

      Motion Time (total) Previous Cycle Time Hour minute second  \
0             00273:58:52         00000:02:39           15:17:01   
1             00273:58:52         00000:02:39           15:17:02   
2             00273:58:52         00000:02:39           15:17:02   
3             00273:58:52         00000:02:39           15:17:03   
4             00273:58:52         00000:02:39           15:17:03   
...                   ...                 ...                ...   
19378         00273:58:55         00000:02:39           13:29:40   
19379         00273:58:55         00000:02:39           13:29:41   
19380         00273:58:55         00000:02:39           13:29:41   
19381         00273:58:55         00000:02:39           13:29:42   
19382         00273:58:55         00000:02:39           13:29:42   

       Tool Changes (total)  ... Present part timer (read only)  \
0                      2975  ...                          159.0   
1                      2975  ...                          159.0   
2                      2975  ...                          159.0   
3                      2975  ...                          159.0   
4                      2975  ...                          159.0   
...                     ...  ...                            ...   
19378                  2975  ...                          159.0   
19379                  2975  ...                          159.0   
19380                  2975  ...                          159.0   
19381                  2975  ...                          159.0   
19382                  2975  ...                          159.0   

       Previous part timer (read only)  Last Block (Modal) Group Codes 1  \
0                                159.0                               0.0   
1                                159.0                               0.0   
2                                159.0                               0.0   
3                                159.0                               0.0   
4                                159.0                               0.0   
...                                ...                               ...   
19378                            159.0                               0.0   
19379                            159.0                               0.0   
19380      

In [23]:
df["Max axis load for X"].value_counts()

Max axis load for X
 0.0    9695
-1.0    9688
Name: count, dtype: int64

In [24]:
df["Max axis load for Y"].value_counts()

Max axis load for Y
50.0    12316
23.0     7067
Name: count, dtype: int64

In [25]:
df["Max axis load for Z"].value_counts()

Max axis load for Z
 0.0    9697
-1.0    9686
Name: count, dtype: int64

In [26]:
df["Motion Time (total)"].value_counts()

Motion Time (total)
00273:58:55    12316
00273:58:52     7067
Name: count, dtype: int64

In [27]:
mod_df["Motion Time (total)"] = pd.to_timedelta(mod_df["Motion Time (total)"])

In [28]:
df["Previous Cycle Time"].value_counts()

Previous Cycle Time
00000:02:39    19383
Name: count, dtype: int64

In [29]:
df["Max recorded vibrations of tool 10"].value_counts()

Max recorded vibrations of tool 10
1.519993    19383
Name: count, dtype: int64

In [30]:
mod_df.drop(columns=["Previous Cycle Time", "Max recorded vibrations of tool 10"], inplace=True)

In [31]:
mod_df.head()

Timestamp Year month day  Cycle start timer  \
0 2025-03-03 15:11:06.163873     2025/03/03           986332.0   
1 2025-03-03 15:11:06.209764     2025/03/03           986332.0   
2 2025-03-03 15:11:06.335765     2025/03/03           986332.0   
3 2025-03-03 15:11:07.035919     2025/03/03           986332.0   
4 2025-03-03 15:11:07.160275     2025/03/03           986332.0   

   Max axis load for X  Max axis load for Y  Max axis load for Z  \
0                  0.0                 23.0                  0.0   
1                 -1.0                 23.0                 -1.0   
2                  0.0                 23.0                  0.0   
3                 -1.0                 23.0                 -1.0   
4                  0.0                 23.0                  0.0   

  Motion Time (total) Hour minute second  Tool Changes (total)  \
0    11 days 09:58:52           15:17:01                  2975   
1    11 days 09:58:52           15:17:02                  2975   
2    11 days 09:58:52           15:17:02                  2975   
3    11 days 09:58:52           15:17:03                  2975   
4    11 days 09:58:52           15:17:03                  2975   

  Power-on Time (total)  ...  Tool in spindle (read only)  \
0           04430:56:44  ...                         10.0   
1           04430:56:45  ...                         10.0   
2           04430:56:45  ...                         10.0   
3           04430:56:45  ...                         10.0   
4           04430:56:46  ...                         10.0   

   Present part timer (read only) Previous part timer (read only)  \
0                           159.0                           159.0   
1                           159.0                           159.0   
2                           159.0                           159.0   
3                           159.0                           159.0   
4                           159.0                           159.0   

   Last Block (Modal) Group Codes 1  Last Block (Modal) Group Codes 2  \
0                               0.0                              17.0   
1                               0.0                              17.0   
2                               0.0                              17.0   
3                               0.0                              17.0   
4                               0.0                              17.0   

   Last complete part timer (read only)  \
0                                 159.0   
1                                 159.0   
2                                 159.0   
3                                 159.0   
4                                 159.0   

   M30 Parts Counter #1 (resettable at control)  \
0                                          2101   
1                                          2101   
2                                          2101   
3                                          2101   
4                                          2101   

   M30 Parts Counter #2 (resettable at control)  \
0                                          2101   
1                                          2101   
2                                          2101   
3                                          2101   
4                                          2101   

   Spindle load with Haas vector drive (read only)  \
0                                              0.0   
1                                              0.0   
2                                              0.0   
3                                              0.0   
4                                              0.0   

   Three-in-one (PROGRAM Oxxxxx STATUS PARTS xxxxx)  
0                   PROGRAM, MDI, IDLE, PARTS, 2101  
1                   PROGRAM, MDI, IDLE, PARTS, 2101  
2                   PROGRAM, MDI, IDLE, PARTS, 2101  
3                   PROGRAM, MDI, IDLE, PARTS, 2101  
4                   PROGRAM, MDI, IDLE, PARTS, 2101  

[5 rows x 23 columns]

In [32]:
df["Hour minute second"].value_counts()

Hour minute second
12:25:57    6
12:44:37    6
16:02:26    6
12:53:02    6
16:02:20    6
           ..
13:12:23    1
13:12:22    1
16:00:36    1
13:11:56    1
15:17:01    1
Name: count, Length: 7048, dtype: int64

In [33]:
df["Tool Changes (total)"].value_counts()

Tool Changes (total)
2975    19383
Name: count, dtype: int64

In [34]:
df["Power-on Time (total)"].value_counts()

Power-on Time (total)
04458:24:19    6
04458:24:53    6
04458:27:42    6
04458:29:03    6
04459:23:31    6
              ..
04431:36:00    1
04431:35:53    1
04431:35:43    1
04431:35:41    1
04430:56:44    1
Name: count, Length: 7059, dtype: int64

In [35]:
mod_df["Power-on Time (total)"] = pd.to_timedelta(mod_df["Power-on Time (total)"])

In [36]:
df["Spindle RPM (read only)"].value_counts()

Spindle RPM (read only)
0    19383
Name: count, dtype: int64

In [37]:
df["Power on timer (read only)"].value_counts()

Power on timer (read only)
15840091.0    6
15941209.0    6
15942319.0    6
15840130.0    6
15940445.0    6
             ..
15842662.0    1
15842666.0    1
15942109.0    1
15940916.0    1
15839945.0    1
Name: count, Length: 7052, dtype: int64

In [38]:
df["Mode (LIST PROG MDI etc.)"].value_counts()

Mode (LIST PROG MDI etc.)
ZERO    19383
Name: count, dtype: int64

In [39]:
df["Tool in spindle (read only)"].value_counts()

Tool in spindle (read only)
10.0    19383
Name: count, dtype: int64

In [40]:
mod_df.drop(columns=["Spindle RPM (read only)", "Tool Changes (total)", "Mode (LIST PROG MDI etc.)", "Tool in spindle (read only)"], inplace=True)
mod_df.head()

Timestamp Year month day  Cycle start timer  \
0 2025-03-03 15:11:06.163873     2025/03/03           986332.0   
1 2025-03-03 15:11:06.209764     2025/03/03           986332.0   
2 2025-03-03 15:11:06.335765     2025/03/03           986332.0   
3 2025-03-03 15:11:07.035919     2025/03/03           986332.0   
4 2025-03-03 15:11:07.160275     2025/03/03           986332.0   

   Max axis load for X  Max axis load for Y  Max axis load for Z  \
0                  0.0                 23.0                  0.0   
1                 -1.0                 23.0                 -1.0   
2                  0.0                 23.0                  0.0   
3                 -1.0                 23.0                 -1.0   
4                  0.0                 23.0                  0.0   

  Motion Time (total) Hour minute second Power-on Time (total)  \
0    11 days 09:58:52           15:17:01     184 days 14:56:44   
1    11 days 09:58:52           15:17:02     184 days 14:56:45   
2    11 days 09:58:52           15:17:02     184 days 14:56:45   
3    11 days 09:58:52           15:17:03     184 days 14:56:45   
4    11 days 09:58:52           15:17:03     184 days 14:56:46   

   Power on timer (read only)  Present part timer (read only)  \
0                  15839945.0                           159.0   
1                  15839946.0                           159.0   
2                  15839946.0                           159.0   
3                  15839946.0                           159.0   
4                  15839946.0                           159.0   

   Previous part timer (read only)  Last Block (Modal) Group Codes 1  \
0                            159.0                               0.0   
1                            159.0                               0.0   
2                            159.0                               0.0   
3                            159.0                               0.0   
4                            159.0                               0.0   

   Last Block (Modal) Group Codes 2  Last complete part timer (read only)  \
0                              17.0                                 159.0   
1                              17.0                                 159.0   
2                              17.0                                 159.0   
3                              17.0                                 159.0   
4                              17.0                                 159.0   

   M30 Parts Counter #1 (resettable at control)  \
0                                          2101   
1                                          2101   
2                                          2101   
3                                          2101   
4                                          2101   

   M30 Parts Counter #2 (resettable at control)  \
0                                          2101   
1                                          2101   
2                                          2101   
3                                          2101   
4                                          2101   

   Spindle load with Haas vector drive (read only)  \
0                                              0.0   
1                                              0.0   
2                                              0.0   
3                                              0.0   
4                                              0.0   

  Three-in-one (PROGRAM Oxxxxx STATUS PARTS xxxxx)  
0                  PROGRAM, MDI, IDLE, PARTS, 2101  
1                  PROGRAM, MDI, IDLE, PARTS, 2101  
2                  PROGRAM, MDI, IDLE, PARTS, 2101  
3                  PROGRAM, MDI, IDLE, PARTS, 2101  
4                  PROGRAM, MDI, IDLE, PARTS, 2101

In [41]:
df["Present part timer (read only)"].value_counts()

Present part timer (read only)
159.0    19383
Name: count, dtype: int64

In [42]:
df["Previous part timer (read only)"].value_counts()

Previous part timer (read only)
159.0    19383
Name: count, dtype: int64

In [43]:
df["Last Block (Modal) Group Codes 1"].value_counts()

Last Block (Modal) Group Codes 1
0.0    19383
Name: count, dtype: int64

In [44]:
df["Last Block (Modal) Group Codes 2"].value_counts()

Last Block (Modal) Group Codes 2
17.0    19383
Name: count, dtype: int64

In [45]:
df["Last complete part timer (read only)"].value_counts()

Last complete part timer (read only)
159.0    19383
Name: count, dtype: int64

In [46]:
df["M30 Parts Counter #1 (resettable at control)"].value_counts()

M30 Parts Counter #1 (resettable at control)
2101    19383
Name: count, dtype: int64

In [47]:
df["M30 Parts Counter #2 (resettable at control)"].value_counts()

M30 Parts Counter #2 (resettable at control)
2101    19383
Name: count, dtype: int64

In [48]:
df["Spindle load with Haas vector drive (read only)"].value_counts()

Spindle load with Haas vector drive (read only)
0.0    19383
Name: count, dtype: int64

In [49]:
df["Three-in-one (PROGRAM Oxxxxx STATUS PARTS xxxxx)"].value_counts()

Three-in-one (PROGRAM Oxxxxx STATUS PARTS xxxxx)
PROGRAM, MDI, IDLE, PARTS, 2101    19383
Name: count, dtype: int64

In [50]:
mod_df.drop(columns=["Present part timer (read only)", "Previous part timer (read only)", "Last Block (Modal) Group Codes 1", "Last Block (Modal) Group Codes 2", "Last complete part timer (read only)", "M30 Parts Counter #1 (resettable at control)", "M30 Parts Counter #2 (resettable at control)", "Spindle load with Haas vector drive (read only)", "Three-in-one (PROGRAM Oxxxxx STATUS PARTS xxxxx)"], inplace=True)
mod_df.head()

Timestamp Year month day  Cycle start timer  \
0 2025-03-03 15:11:06.163873     2025/03/03           986332.0   
1 2025-03-03 15:11:06.209764     2025/03/03           986332.0   
2 2025-03-03 15:11:06.335765     2025/03/03           986332.0   
3 2025-03-03 15:11:07.035919     2025/03/03           986332.0   
4 2025-03-03 15:11:07.160275     2025/03/03           986332.0   

   Max axis load for X  Max axis load for Y  Max axis load for Z  \
0                  0.0                 23.0                  0.0   
1                 -1.0                 23.0                 -1.0   
2                  0.0                 23.0                  0.0   
3                 -1.0                 23.0                 -1.0   
4                  0.0                 23.0                  0.0   

  Motion Time (total) Hour minute second Power-on Time (total)  \
0    11 days 09:58:52           15:17:01     184 days 14:56:44   
1    11 days 09:58:52           15:17:02     184 days 14:56:45   
2    11 days 09:58:52           15:17:02     184 days 14:56:45   
3    11 days 09:58:52           15:17:03     184 days 14:56:45   
4    11 days 09:58:52           15:17:03     184 days 14:56:46   

   Power on timer (read only)  
0                  15839945.0  
1                  15839946.0  
2                  15839946.0  
3                  15839946.0  
4                  15839946.0

In [51]:
df.head(1)

Index                   Timestamp  Feed timer Last Cycle Time  \
0      4  2025-03-03 15:11:06.163873    694348.0     00000:02:39   

  Year month day  Cycle start timer  Tool Number in use  Max axis load for X  \
0     2025/03/03           986332.0                  10                  0.0   

   Max axis load for Y  Max axis load for Z  ...  \
0                 23.0                  0.0  ...   

  Present part timer (read only) Previous part timer (read only)  \
0                          159.0                           159.0   

  Last Block (Modal) Group Codes 1  Last Block (Modal) Group Codes 2  \
0                              0.0                              17.0   

  Max recorded vibrations of tool 10  Last complete part timer (read only)  \
0                           1.519993                                 159.0   

   M30 Parts Counter #1 (resettable at control)  \
0                                          2101   

  M30 Parts Counter #2 (resettable at control)  \
0                                         2101   

   Spindle load with Haas vector drive (read only)  \
0                                              0.0   

   Three-in-one (PROGRAM Oxxxxx STATUS PARTS xxxxx)  
0                   PROGRAM, MDI, IDLE, PARTS, 2101  

[1 rows x 29 columns]

There are plenty of features in this dataset, many of which consist of the same values.
- Timestamp
- Index (deleted)
- Year month day (part of timestamp)
- Hour minute second (part of timestamp, slightly differs from timestamp's time)
- Feed timer: 694348 (deleted)
- Last cycle time: 02:39 (deleted)
- Cycle start timer: 986335 or 986332
- Tool Number in use: 10 (deleted)
- Max axis load for X: 0 or -1
- Max axis load for Y: 50 or 23
- Max axis load for Z: 0 or -1
- Motion Time (total): 273:58:55 or 273:58:52
- Previous Cycle Time: 02:39 (deleted)
- Tool Changes (total): 2975 (deleted)
- Spindle RPM (read only): 0 (deleted)
- Mode (LIST PROG MDI etc.): ZERO (deleted)
- Tool in spindle (read only): 10 (deleted)
- Present part timer (read only): 159 (deleted)
- Previous part timer (read only): 159 (deleted)
- Last Block (Modal) Group Codes 1: 0.0 (deleted)
- Last Block (Modal) Group Codes 2: 17.0 (deleted)
- Max recorded vibrations of Tool 10: 1.5199931014055006 (deleted)
- Last complete part timer (read only): 159.0 (deleted)
- M30 Parts Counter #1-2 (ressettable at control): 2101 (deleted)
- Spindle load with Haas vector drive (read only): 0.0 (deleted)
- Three-in-one (PROGRAM 0xxxxx STATUS PARTS xxxxx): PROGRAM, MDI, IDLE, PARTS, 2101 (deleted)

In [52]:
mod_df["Max axis load for (X, Y, Z)"] = df.apply(lambda row: f"({row['Max axis load for X']}, {row['Max axis load for Y']}, {row['Max axis load for Z']})", axis=1)

In [53]:
mod_df["Max axis load for (X, Y, Z)"].value_counts()

Max axis load for (X, Y, Z)
(0.0, 50.0, 0.0)      6153
(-1.0, 50.0, -1.0)    6145
(0.0, 23.0, 0.0)      3531
(-1.0, 23.0, -1.0)    3530
(-1.0, 50.0, 0.0)       12
(0.0, 50.0, -1.0)        6
(0.0, 23.0, -1.0)        5
(-1.0, 23.0, 0.0)        1
Name: count, dtype: int64

Data Visualizations

In [54]:
report = sweetviz.analyze(mod_df)
report.show_html()

NameError: name 'sweetviz' is not defined

In [55]:
profile = ProfileReport(mod_df, title="Original Dataset Profile")
profile.to_file('YDATA_PROFILE.html')

Summarize dataset:   0%|          | 0/5 [00:00<?, ?it/s]


100%|██████████| 11/11 [00:00<00:00, 70.74it/s][A
/opt/homebrew/lib/python3.11/site-packages/ydata_profiling/model/pandas/discretize_pandas.py:52: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[0 0 0 ... 9 9 9]' has dtype incompatible with timedelta64[ns], please explicitly cast to a compatible dtype first.
  discretized_df.loc[:, column] = self._discretize_column(
/opt/homebrew/lib/python3.11/site-packages/ydata_profiling/model/pandas/discretize_pandas.py:52: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value '[0 0 0 ... 9 9 9]' has dtype incompatible with timedelta64[ns], please explicitly cast to a compatible dtype first.
  discretized_df.loc[:, column] = self._discretize_column(


Generate report structure:   0%|          | 0/1 [00:00<?, ?it/s]

Render HTML:   0%|          | 0/1 [00:00<?, ?it/s]

Export report to file:   0%|          | 0/1 [00:00<?, ?it/s]